prereise/gather/winddata/hrrr/calculations.py

```diff
@@ -132,3 +162,81 @@ def calculate_pout(wind_farms, start_dt, end_dt, directory):
     df = pd.DataFrame(data=wind_power_data, index=dts, columns=wind_farms.index)
 
     return df
+
+
+def calculate_pout_individual(wind_farms, start_dt, end_dt, directory):
+    """Calculate power output for wind farms based on hrrr data. Each wind farm's power
+    curve is based on farm-specific attributes.
+    Function assumes that user has already called
+    :meth:`prereise.gather.winddata.hrrr.hrrr.retrieve_data` with the same
+    start_dt, end_dt, and directory.
+
+    :param pandas.DataFrame wind_farms: plant data frame, plus additional columns:
+        'Predominant Turbine Manufacturer', 'Predominant Turbine Model Number', and
+        'Turbine Hub Height (Feet)'.
+    :param str start_dt: start date.
+    :param str end_dt: end date (inclusive).
+    :param str directory: directory where hrrr data is contained.
+    :return: (*pandas.Dataframe*) -- Pandas data frame containing power out per wind
+        farm on a per hourly basis between start_dt and end_dt inclusive. Structure of
+        dataframe is:
+            wind_farm1  wind_farm2
+        dt1    POUT        POUT
+        dt2    POUT        POUT
+    :raises ValueError: if ``wind_farms`` is missing the 'state_abv' column.
+    """
+
+    def get_shifted_curve(wind_farm_series, reference_curves, curve_cache):
+        """Get the power curve for the given turbine type, shifted to hub height. If the
+        turbine curve isn't found in ``reference_curves``, default to IEC class 2. Cache
+        results from identical queries.
+
+        :return: (*pandas.Series*) -- index is wind speed at 80m, values are normalized
+        power output.
+        """
+        full_name = " ".join(
+            [wind_farm_series[const.mfg_col], wind_farm_series[const.model_col]]
+        )
+        hub_height = wind_farm_series[const.hub_height_col]
+        lookup_key = (full_name, hub_height)
+        try:
+            return curve_cache[lookup_key]
+        except KeyError:
+            try:
+                starting_curve = reference_curves[full_name]
+            except KeyError:
+                starting_curve = reference_curves["IEC class 2"]
+            shifted_curve = shift_turbine_curve(
+                starting_curve, hub_height, const.max_wind_speed, const.new_curve_res
+            )
+            curve_cache[lookup_key] = shifted_curve
+            return shifted_curve
```

Comment:
Not sure whether using the built-in cache decorator gives us same behavior here:
```python
@cache
def get_shifted_curve(wind_farm_series, reference_curves):
        full_name = " ".join(
            [wind_farm_series[const.mfg_col], wind_farm_series[const.model_col]]
        )
        hub_height = wind_farm_series[const.hub_height_col]
        starting_curve = reference_curves[full_name] if full name in reference_curves else reference_curves["IEC class 2"]
        shifted_curve = shift_turbine_curve(
            starting_curve, hub_height, const.max_wind_speed, const.new_curve_res
        )
        return shifted_curve
```

Reply:
`@functools.cache`? TIL!

Unfortunately, pandas Series and DataFrames aren't hashable. Python doesn't seem to have a convenient hashable mapping class (although they're considering one in PEP 603), so if we want to pass the same inputs, I think we'd need to mock one up using tuples, and that might end up looking just as ugly as what we have currently.

Reply:
Yeah, I presume `wind_farm_series` and `reference_curves` are either pandas series or data frame. One potential work around would be make `get_shifted_curve` take three parameters instead, i.e. full_name_first_part, full_name_second_part and height, then treat reference_curves as defined variables outside the function, i.e. use `turbine_power_curves` directly. Just random thoughts. I'm okay with the current implementation as well.

Reply:
Another implementation of the same sort of thing would be using `functools.partial` to curry the unchanging reference curves into a function that only takes hashable inputs, so we don't need to rely on anything outside of local scope, and then calling the decorator function directly on the curried function. Something like:

```python
import functools

def calculate_pout_individual(wind_farms, start_dt, end_dt, directory):
    ....
    def get_shifted_curve(full_name, hub_height, reference_curves):
        try:
            starting_curve = reference_curves[full_name]
        except KeyError:
            starting_curve = reference_curves["IEC class 2"]
        return shift_turbine_curve(
            starting_curve, hub_height, const.max_wind_speed, const.new_curve_res
        )

    turbine_power_curves = get_turbine_power_curves()
    partial_func = functools.partial(get_shifted_curve, reference_curves=turbine_power_curves)
    cached_func = functools.cache(partial_func)
    full_name = " ".join(
        [wind_farm_series[const.mfg_col], wind_farm_series[const.model_col]]
    )
    hub_height = wind_farm_series[const.hub_height_col]
    shifted_power_curves = wind_farms.apply(
        lambda x: cached_func(f"{x[const.mfg_col]} {x[const.model_col]}", x[const.hub_height_col]),
        axis=1,
    )
    ...
```

Reply:
TIL!

Reply:
Or maybe to be a bit clearer:

```python
import functools

def calculate_pout_individual(wind_farms, start_dt, end_dt, directory):
    ....
    def get_starting_curve_name(mfg_name, model_name, valid_names):
        full_name = " ".join([mfg_name, model_name])
        return full name if full_name in valid_names else "IEC class 2"

    def get_shifted_curve(lookup_name, hub_height, reference_curves):
        return shift_turbine_curve(
            reference_curves[lookup_name],
            hub_height,
            const.max_wind_speed,
            const.new_curve_res,
        )

    # Create cached, curried function for use with apply
    turbine_power_curves = get_turbine_power_curves()
    partial_func = functools.partial(get_shifted_curve, reference_curves=turbine_power_curves)
    cached_func = functools.cache(partial_func)
    # Create dataframe of lookup values
    lookup_names = wind_farms.apply(lambda x: f"{x[const.mfg_col]} {x[const.model_col]}", axis=1)
    lookup_values = pandas.concat([lookup_names, wind_farms[const.hub_height_col]], axis=1)
    # Use lookup values with cached, curried function
    shifted_power_curves = lookup_values.apply(lambda x: cached_func(*x), axis=1)
    ...
```

Reply:
One final note: `@functools.cache(myfunc)` isn't available until Python 3.9, so for python 3.8 backward-compatibility we need `@functools.lru_cache(maxsize=None)(myfunc)` instead.

Reply:
Yeah, I should have put this note down as well. The `cache` decorator is something I learned when my compiler is complaining that `@lru_cache` needs to specify a size parameter or call as a function, i.e. `@lru_cache()`.